homeassistant/components/turn_touch.py

```diff
@@ -0,0 +1,189 @@
+        hass.data[DATA_KEY]['devices'][turn_touch.address] = turn_touch
+
+    for platform in TURN_TOUCH_PLATFORMS:
+        load_platform(hass, platform, DOMAIN, {}, config)
+
+    return True
+
+
+class TurnTouchRemote:
+    """Representation of a Turn Touch remote.
+
+    This does not inherit from the Entity class because Entities are created
+    by platforms (e.g., sensor.turn_touch). But we must maintain one shared
+    connection to the device, so that happens here.
+    """
+
+    def __init__(self, hass, config, listen=True):
+        """Initialize the remote."""
+        # pylint: disable=import-error
+        import turntouch
+        self._hass = hass
+        self.address = config[CONF_MAC]
+        self._debounce = config[CONF_DEBOUNCE]
+        self._device = None
+
+        class Handler(turntouch.DefaultActionHandler):
+            """Callback handler for button presses."""
+
+            def __init__(self, handler_func):
+                self._handler_func = handler_func
+
+            def action_any(self, action: turntouch.Action = None):
+                """Handle all button presses."""
+                if not action.is_off:
+                    self._handler_func(action.name)
+
+        self._handler = Handler(self._button_press)
+        self.name = DEFAULT_NAME
+        self._connect_lock = threading.Lock()
+        self._io_lock = threading.Lock()
+        threading.Thread(target=self._setup).start()
+
+    def _setup(self):
+        """Open the connection to the remote and start listening for events."""
+        self._connect()
+        self.get_name()
+        self._listen()
+
+    def get_name(self):
+        """Read the device name."""
+        # pylint: disable=import-error
+        import turntouch
+        delay = RETRY_DELAY
+        while True:
+            try:
+                with self._io_lock:
+                    self.name = self._device.name
+                return self.name
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Reading device name failed. Retrying...')
+            except AttributeError:
+                _LOGGER.warning('Cannot read name before connecting'
+                                ' to the device. Retrying...')
+            time.sleep(delay)
+            delay *= 2
+            self._connect()
+
+    def get_battery(self):
+        """Read the device battery level."""
+        # pylint: disable=import-error
+        import turntouch
+        delay = RETRY_DELAY
+        while True:
+            try:
+                with self._io_lock:
+                    return self._device.battery
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Reading device battery failed. Retrying...')
+            except AttributeError:
+                _LOGGER.warning('Cannot read battery level before connecting'
+                                ' to the device. Retrying...')
+            time.sleep(delay)
```

Comment:
`get_battery` is still called from the entity `update` method, so sleeping here is not ok.

Reply:
Thanks for your patience and helpful feedback. `update` now kicks off a new thread for the `get_battery` call. I've also moved the `io_lock` (thread safety) stuff into the `TurnTouch` library, since that isn't a Home Assistant problem.